server/game/core/gameSteps/prompts/MulliganPrompt.ts

```diff
@@ -0,0 +1,57 @@
+import { AllPlayerPrompt } from './AllPlayerPrompt';
+import { IPlayerPromptStateProperties } from '../../PlayerPromptState';
+import type Game from '../../Game';
+import * as Contract from '../../utils/Contract';
+
+export class MulliganPrompt extends AllPlayerPrompt {
+    protected playersDone = new Map<string, boolean>();
+
+    public constructor(game: Game) {
+        super(game);
+        for (const player of game.getPlayers()) {
+            this.playersDone[player.name] = false;
+        }
+    }
+
+    public override completionCondition(player): boolean {
+        return this.playersDone[player.name];
+    }
+
+    public override activePrompt(): IPlayerPromptStateProperties {
+        return {
+            menuTitle: 'Do you want to mulligan your hand?',
+            buttons: [{ text: 'Yes', arg: 'yes' }, { text: 'No', arg: 'no' }],
+            promptTitle: 'Mulligan Step',
+            promptUuid: this.uuid
+        };
+    }
+
+    public override waitingPrompt() {
+        return {
+            menuTitle: 'Waiting for opponent to choose whether to Mulligan or keep hand.'
+        };
+    }
+
+    public override menuCommand(player, arg): boolean {
+        if (arg === 'yes') {
+            if (this.completionCondition(player)) {
+                return false;
+            }
+            for (const card of player.hand) {
+                player.moveCard(card, 'deck bottom');
+            }
+
+            player.shuffleDeck();
```

Comment:
Similar to the resourcing, we should wait until both players have committed to their choice before doing anything. This will prevent people from trying to base their choice on whether the other person mulligans

Reply:
You got it!